helix-view/src/view.rs

```diff
@@ -161,6 +161,62 @@ impl View {
         Some(Position::new(row, col))
     }
 
+    /// Verifies whether a screen position is inside the view
+    /// Returns true when position is inside the view
+    pub fn verify_screen_coords(&self, row: usize, column: usize) -> bool {
+        // 2 for status
+        if row < self.area.y as usize || row > self.area.y as usize + self.area.height as usize - 2
+        {
+            return false;
+        }
+
+        // TODO: not ideal
+        const OFFSET: usize = 7; // 1 diagnostic + 5 linenr + 1 gutter
+
+        if column < self.area.x as usize + OFFSET
+            || column > self.area.x as usize + self.area.width as usize
+        {
+            return false;
+        }
+        true
+    }
+
+    /// Translates a screen position to position in the text document.
+    /// Returns a usize typed position in bounds of the text if found in this view, None if out of view.
+    pub fn pos_at_screen_coords(&self, doc: &Document, row: usize, column: usize) -> Option<usize> {
+        if !self.verify_screen_coords(row, column) {
+            return None;
+        }
+
+        let text = doc.text();
+        let line_number = row - self.area.y as usize + self.first_line;
+        let mut pos = text.line_to_char(line_number);
+
+        let current_line = text.line(line_number);
+        let tab_width = doc.tab_width();
+
+        // TODO: not ideal
+        const OFFSET: usize = 7; // 1 diagnostic + 5 linenr + 1 gutter
+
+        let target = column - OFFSET - self.area.x as usize + self.first_col;
+        let mut selected = 0;
+
+        for grapheme in RopeGraphemes::new(current_line) {
+            if selected > target {
+                break;
+            }
+            if grapheme == "\t" {
+                selected += tab_width;
+            } else if grapheme != "\n" {
+                let width = grapheme_width(&Cow::from(grapheme));
+                selected += width;
+            }
+            pos += 1;
+        }
+
+        Some(pos - 1)
```

Comment:
Needs to be a saturating sub probably otherwise clicking on (0, 0) will panic, also why the `- 1`?

Reply:
User clicks after the line end, all the length gets counted and so overflowed to the next line. No panic happens with (0, 0)

Reply:
I mean if I click on the top left first character of the first line. `pos` there will be 0 I think?

Reply:
Looks like not, since nothing crashes. Anyway, it'd be nice to count whole line lenght - 1 with the loop iterating over graphemes.

Reply:
Along with the comments I made above, I think what we want instead of `Some(pos - 1)` here is `Some(pos.min(line_end_char_index(text.slice(..), line_number)))`.  Basically, just ensuring that it doesn't go off the end of the line.  (You can find `line_end_char_index()` in `helix_core::line_ending`.)

Getting code like this right can be subtle, so I may have missed some things or made a mistake in my suggestions.  I definitely would feel more comfortable about this with unit tests, to make sure we're getting things right.